### PR TITLE
Add codestral mamba2

### PR DESCRIPTION
**Pull request by molbap**
# What does this PR do?

This PR adds mamba2 to support mistral's mamba-codestral release.

- [x] Make sure weights are correctly loaded
- [x] Call correct cuda kernel in case of self.training, like mamba_inner_fn for mamba1
- [x] Cuda forward actually supports outputting states - we can use it for caching
- [x] Implement slow_forward as well
- [x] fix quality of slow forward generations
- [x] remove einops
- [x] Really remove all einops
- [x] Figure out why cuda kernels are actually slow? edit: it's unclear but an issue raised in the original repo https://github.com/state-spaces/mamba/issues/389#issuecomment-2171755306
- [x] add tests
- [x] Fix batched generation by tuning out states (thanks @vasqu !), edit: pointed out in https://github.com/state-spaces/mamba/issues/66#issuecomment-1863563829, affects mamba-1 as well likely
- [x] add more integration tests (batched generation)

cc @ArthurZucker

### CODE REVIEW

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 1]
```diff
@@ -0,0 +1,762 @@
...
+    add_start_docstrings_to_model_forward,
+    logging,
+)
+from ...utils.import_utils import is_causal_conv1d_available, is_mamba_ssm_available
+from .configuration_mamba2 import Mamba2Config
+
+
+logger = logging.get_logger(__name__)
+
+if is_mamba_ssm_available():
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:41]: Should this not be a version check since `mamba_ssm` only provides the necessary functions after v2.0 and there has been a bug on first release that caused NaN losses which has been fixed with [>=2.0.4](https://github.com/state-spaces/mamba/issues/352#issuecomment-2162223186). 

Further [info](https://github.com/state-spaces/mamba/issues/352#issuecomment-2167093395) that might be provided in the docs: There is a restriction on the parameters with `d_model * expand / headdim = multiple of 8`.

**Review by vasqu — COMMENTED** [review]
Hope you don't mind a few comments of mine. Good luck on the implementation, looking forward to the release!

**Comment by Adibvafa** [comment]
I believe it will be more efficient to build on top of https://github.com/huggingface/transformers/pull/32027 to speed up the integration process!

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 2]
```diff
@@ -0,0 +1,762 @@
...
+            y = y.view(y.shape[0], y.shape[1], -1)
+            # Multiply "gate" branch and apply extra normalization layer
+
+            y = self.norm(y, gate)
+            out = self.out_proj(y)
+        return out
+
+    # fmt: off
+    # FIXME slow generations are lower quality
+    def slow_forward(self, input_states, cache_params: Optional[Mamba2Cache]=None):
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/mamba2/modeling_mamba2.py:277]: Yeah but the variant does not work for arbitrary length so we left it on the side for now!

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/mamba2/modeling_mamba2.py:277]: We'll checkout what you did!

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/mamba2/modeling_mamba2.py:277]: (no single letter variables pleasssssse 👿 )

[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/mamba2/modeling_mamba2.py:277]: Hard to do when it's referring to the SSM variables :sweat: the inner functions could probably be renamed instead of x I guess.

**Comment by ArthurZucker** [comment]
Hey both! We tried your code and could not get it to work, mostly the shape of the output final_state is wrong (ends with 1024 vs 128 that we need). We are probably gonna ship it as is, but a patch will be possible if we can ship the ssimd in it!

**Comment by vasqu** [comment]
@ArthurZucker I've just seen that it's using groups, so the code doesn't work out of the box as it would need appropriate resizing of B and C for example (ngroups -> nheads by repetition). I did not account for that as the original models never used anything other than group=1. Just tested it locally and it seems to work with those slight adjustments on B and C.  

I can't check in on the final_state shape without seeing the code so maybe I can help later when you guys ship it and there is a base to work on :D

**Comment by molbap** [comment]
Hey both again, thanks for the comment @vasqu ! indeed it seems that it needs to be slightly modified - a `B.repeat(1, 1, self.num_heads // self.num_groups, 1)` seems promising. Are you verifying with a .generate() what the model is outputting? We'll ship this soon as Arthur mentioned and credit your code/include you as coauthors of course :hugs:

**Comment by vasqu** [comment]
@molbap I've been using my [mamba2-torch](https://github.com/vasqu/mamba2-torch) repo to check modifications on the original `ssd_minimal` script to still work correctly (i.e. with variable seq_len, D skip connection, and just now with the n_groups >= 1). You can check it out over [here](https://github.com/vasqu/mamba2-torch/blob/main/tests/TestSSDMinimal.py). It does use einops though which I've changed in our PR to be pure torch.

**Comment by molbap** [comment]
I changed up generations - the cuda kernel calls + the naive call have slightly different generations but both are very reasonable and seem similar to the original implementation. Missing some tests + remove the triple einsum contraction to create Ydiag and should be good to go!

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 3]
```diff
@@ -0,0 +1,1009 @@
...
+
+        # 2. Compute the state for each intra-chunk
+        # (right term of low-rank factorization of off-diagonal blocks; B terms)
+        decay_states = torch.exp((A_cumsum[:, :, :, -1:] - A_cumsum))
+        states = torch.einsum("bclhn,bhcl,bclhp->bchpn", B, decay_states, hidden_states)
+
+        # 3. Compute the inter-chunk SSM recurrence; produces correct SSM states at chunk boundaries
+        # (middle term of factorization of off-diag blocks; A terms)
+        if cache_params is not None and cache_params.seqlen_offset > 0:
+            previous_states = cache_params.ssm_states[self.layer_idx].unsqueeze(1)
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:556]: Small suggestion on the generation via caching as we can avoid the following computations / simplify them just as in the fast path, see:
https://github.com/huggingface/transformers/blob/7e148149a1c809065e7ca720d170443c50bb7d69/src/transformers/models/mamba2/modeling_mamba2.py#L290-L327 

The equivalent for the slow path would be:
https://github.com/huggingface/transformers/blob/2e44073e4913b6ca26dbf7e0003cc07d36b0dd34/src/transformers/models/mamba2/modeling_mamba2.py#L1184-L1208

Maybe that helps even further in improving generation quality?

Comment (vasqu) [reply src/transformers/models/mamba2/modeling_mamba2.py:556]: Mb, it needs to be adjusted for groups again. This should work (not thoroughly tested though):
```python
    # Get time step with softplus and bias
    # [bsz, 1, num_heads] -> [bsz, num_heads, head_dim]
    dt = dt.transpose(1, 2).expand(dt.shape[0], dt.shape[-1], self.head_dim)
    # [num_heads] -> [num_heads, head_dim]
    dt_bias = dt_bias.unsqueeze(-1).expand(dt_bias.shape[0], self.head_dim)
    dt = torch.nn.functional.softplus(dt + dt_bias.to(dt.dtype))
    
    
    # Discretize A
    # [num_heads] -> [num_heads, head_dim, state_size]
    A = A.unsqueeze(-1).unsqueeze(-1).expand(A.shape[0], self.head_dim, self.ssm_state_size).to(dtype=torch.float32)
    # [bsz, num_heads, head_dim, state_size]
    dA = torch.exp(dt.unsqueeze(-1) * A)
    
    # Discretize B
    # [bsz, n_groups * state_size] -> [bsz, n_groups, 1, state_size] ->
    # -> [bsz, n_groups, group to head repetition factor, state_size] -> [bsz, num_heads, state_size]
    B = B.reshape(B.shape[0], self.n_groups, -1).unsqueeze(-2)
    B = B.expand(B.shape[0], B.shape[1], self.num_heads // self.n_groups, B.shape[-1]).contiguous()
    B = B.reshape(B.shape[0], -1, B.shape[-1])
    # [bsz, num_heads, head_dim, state_size]
    dB = dt.unsqueeze(-1) * B.unsqueeze(-2)
    
    # Discretize x into dB
    # [bsz, intermediate_size] -> [bsz, num_heads, head_dim]
    x = x.reshape(x.shape[0], -1, self.head_dim)
    dBx = dB * x.unsqueeze(-1)
    
    # State calculation
    cache_params.ssm_states[self.layer_idx].copy_(
        cache_params.ssm_states[self.layer_idx] * dA + dBx
    )
    
    # Subsequent output
    # [bsz, n_groups * state_size] -> [bsz, num_heads, state_size]
    C = C.reshape(C.shape[0], self.n_groups, -1).unsqueeze(-2)
    C = C.expand(C.shape[0], C.shape[1], self.num_heads // self.n_groups, C.shape[-1]).contiguous()
    C = C.reshape(C.shape[0], -1, C.shape[-1])
    # [bsz, num_heads, head_dim]
    y = torch.einsum("bhdn,bhn->bhd", cache_params.ssm_states[self.layer_idx].to(C.dtype), C)
    
    # D skip connection
    # [num_heads] -> [num_heads, head_dim]
    D = D.unsqueeze(-1).expand(D.shape[0], self.head_dim)
    y = y + x * D.unsqueeze(0)
    
    # [bsz, num_heads, head_dim] -> [bsz, 1, intermediate_size]
    y = y.reshape(y.shape[0], -1).unsqueeze(1)
```

Comment (molbap) [reply src/transformers/models/mamba2/modeling_mamba2.py:556]: Yes, hadn't finished up the cached part entirely - I think your reshapes are correct, missing the dt clamping after softplus maybe?

Comment (vasqu) [reply src/transformers/models/mamba2/modeling_mamba2.py:556]: Yes, totally forgot about that 👍

**Comment by molbap** [comment]
Alright, I've added a non-einsum equivalent of the tensor contraction paths for both the diag and off-diag terms that seem to give good results, supports batched inputs as well.

**Comment by molbap** [comment]
@vasqu Haha, unfortunately it didn't it seems - there's still a difference in quality between the two paths. But I think it's reasonable for now, would need to do proper benchmarks.

**Comment by vasqu** [comment]
Just saw the "Figure out why cuda kernels are actually slow?"

I've also wondered about that when i worked on mamba2 and here's an issue of mine I used to keep track: https://github.com/vasqu/mamba2-torch/issues/1#issue-2349175830

It seems like the triton kernels cause a significant cpu overhead on the first iteration(s): https://github.com/state-spaces/mamba/issues/389#issuecomment-2171755306

Other suggestions: big model and/or compiling https://github.com/state-spaces/mamba/issues/355#issuecomment-2147597457 (but that breaks in a lot of cases which happened in my case too)

**Comment by molbap** [comment]
Yes! Been reading that one too and couldn't get the compile to improve performances (ie crashes). The warmup run definitely helps as well. So this full torch path is currently faster.

**Comment by molbap** [comment]
I added some tests and did last touches on modeling - Batched generation is a bit weird, I think there is leaking of states within a batch and not sure this is fixed in the original repo either so keeping it as is.

**Comment by vasqu** [comment]
Late night me, regarding batched generation: Do the padding sides differ much (left vs right)? I'd assume that we have leaky information flowing through either way. This affects both mamba 1+2. 

When using left padding we would pad with the eos token and subsequently with a non-zero input which would influence causal directions (left to right, repeated eos information is given to the SSM and conv1d). In attention a simple attention mask would take care of it for the core softmax operations, for mamba there does not exist such thing yet. I'd suggest using the original attention mask to zero the padding tokens before entering the mamba block (zero input and zero initial hidden state would result in repeated zeros until encountering the first non zero input). It has been similarly suggested back then for the original mamba model: https://github.com/state-spaces/mamba/issues/66#issuecomment-1863563829

Right padding would probably need keeping track of the sequence indices which recent versions of mamba and conv1d support. Without recent versions it would probably get less "elegant".

**Comment by molbap** [comment]
Indeed, that's likely the issue, thanks @vasqu ! I tried tuning out the hidden states but didn't think of doing it in two places :face_with_peeking_eye: It seems that fixes the batched generation, I'll push that soon

**Comment by vasqu** [comment]
@molbap Does that also work for right padding? Reassigning zeros would probably spit out garbage imo (without retrieving the correct last values). Maybe raise a warning and do not interact with the attention mask if right padding?

**Comment by molbap** [comment]
I don't think it does right now for right padding no, but it should work in all other cases now at least!

File: docs/source/en/model_doc/mamba2.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+*<INSERT PAPER ABSTRACT HERE>*
+
+Tips:
+
+<INSERT TIPS ABOUT MODEL HERE>
+
+This model was contributed by [INSERT YOUR HF USERNAME HERE](https://huggingface.co/<INSERT YOUR HF USERNAME HERE>).
+The original code can be found [here](<INSERT LINK TO GITHUB REPO HERE>).
+
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/mamba2.md:36]: this needs filling!

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/mamba2/modeling_mamba2.py:41]: cc @molbap we need to check version here!

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 5]
```diff
@@ -0,0 +1,1123 @@
...
+# Helper methods for segment sum computation
+
+
+def pad_by_size(x, pad_size):
+    """
+    Padding x tensor with `pad_size` on the seq_len dim (dim=1)
+
+    Assumes that we only have tensors of either size 4 or 3
+    """
+    assert 2 < len(x.shape) < 5
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:67]: no asserts please

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 6]
```diff
@@ -0,0 +1,1123 @@
...
+        # b (l c) h -> b l c h with c=chunk_size
+        # [bsz, seq_len multiple of chunk_size, num_heads] -> [bsz, -1, chunk_size, num_heads]
+        return x.reshape(x.shape[0], -1, chunk_size, x.shape[2])
+    else:
+        # b (l c) h p -> b l c h p with c=chunk_size
+        # [bsz, seq_len multiple of chunk_size, num_heads, head_dim or state_size] -> [bsz, -1, chunk_size, num_heads, head_dim or state_size]
+        return x.reshape(x.shape[0], -1, chunk_size, x.shape[2], x.shape[3])
+
+
+def segsum(x):
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:94]: let's avoid single letter variable (valid for all func)

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 7]
```diff
@@ -0,0 +1,1123 @@
...
+    Padding x tensor with `pad_size` on the seq_len dim (dim=1) and
+    simultaneously splitting it into chunk sequences.
+
+    Assumes that we only have tensors of either size 4 or 3
+    """
+    # [bsz, seq_len, ...] -> [bsz, seq_len multiple of chunk_size, ...]
+    x = pad_by_size(x, pad_size)
+
+    if len(x.shape) == 3:
+        # b (l c) h -> b l c h with c=chunk_size
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:85]:
```suggestion
```

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 8]
```diff
@@ -0,0 +1,1123 @@
...
+        # [bsz, seq_len multiple of chunk_size, num_heads] -> [bsz, -1, chunk_size, num_heads]
+        return x.reshape(x.shape[0], -1, chunk_size, x.shape[2])
+    else:
+        # b (l c) h p -> b l c h p with c=chunk_size
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:89]:
```suggestion
```

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 9]
```diff
@@ -0,0 +1,1123 @@
...
+                zxbcdt,
+                [
+                    d_mlp,
+                    d_mlp,
+                    self.intermediate_size,
+                    self.intermediate_size + 2 * self.n_groups * self.ssm_state_size,
+                    self.num_heads,
+                ],
+                dim=-1,
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:291]: let's spslit these as much as possible to make it readable

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 10]
```diff
@@ -0,0 +1,1123 @@
...
+                xBC,
+                [self.intermediate_size, self.n_groups * self.ssm_state_size, self.n_groups * self.ssm_state_size],
+                dim=-1,
+            )
+            A = -torch.exp(self.A_log.float())  # (nheads,)
+
+            A = A.unsqueeze(1).unsqueeze(2).expand(-1, self.head_dim, self.ssm_state_size).to(dtype=torch.float32)
+            dt = dt.unsqueeze(2).expand(-1, -1, self.head_dim)
+            dt_bias = self.dt_bias.unsqueeze(1).expand(-1, self.head_dim)
+            D = self.D.unsqueeze(1).expand(-1, self.head_dim)  # repeat(self.D, "h -> h p", p=self.head_dim)
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:311]: I am more a fan of [:,None,None] notation!

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 11]
```diff
@@ -0,0 +1,1123 @@
...
+                B,
+                C,
+                D,
+                z=gate if not self.rms_norm else None,
+                dt_bias=dt_bias,
+                dt_softplus=True,
+            )
+            hidden_states = hidden_states.view(batch_size, self.num_heads * self.head_dim)
+            if self.rms_norm:
+                hidden_states = self.norm(hidden_states, gate)
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:332]: I think rms norm is always used, let's remove if else

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 12]
```diff
@@ -0,0 +1,1123 @@
...
+            if d_mlp > 0:
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:333]: never 0 here let's remove

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 13]
```diff
@@ -0,0 +1,1123 @@
...
+            dt_limit_kwargs = {} if self.time_step_limit == (0.0, float("inf")) else {"dt_limit": self.time_step_limit}
+
+            if self.training and cache_params is None:
+                out, ssm_state = mamba_split_conv1d_scan_combined(
+                    projected_states,
+                    self.conv1d.weight.squeeze(1),
+                    self.conv1d.bias,
+                    self.dt_bias,
+                    A,
+                    D=self.D.view(-1, self.head_dim) if self.D_has_hdim else self.D,
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:351]: LEt's remove the if else here as well!

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 14]
```diff
@@ -0,0 +1,1123 @@
...
+
+            else:
+                gate, xBC, time_step = torch.split(
+                    projected_states,
+                    [self.intermediate_size, self.conv_dim, self.num_heads],
+                    dim=-1,
+                )
+                if attention_mask is not None and attention_mask.shape[1] > 1 and attention_mask.shape[0] > 1:
+                    # tune out hidden states for pad tokens, see https://github.com/state-spaces/mamba/issues/66
+                    # bug in generate tests?
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:374]: potentially because some test use 4d mask!

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 15]
```diff
@@ -0,0 +1,1123 @@
...
+                hidden_states, B, C = torch.split(
+                    xBC,
+                    [self.intermediate_size, self.n_groups * self.ssm_state_size, self.n_groups * self.ssm_state_size],
+                    dim=-1,
+                )
+
+                if attention_mask is not None and attention_mask.shape[1] > 1 and attention_mask.shape[0] > 1:
+                    # tune out hidden states for pad tokens, see https://github.com/state-spaces/mamba/issues/66
+                    dtype = hidden_states.dtype
+                    hidden_states = (hidden_states * attention_mask.unsqueeze(2)).to(dtype)
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:399]: why are we doing it twice? Interesting

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 16]
```diff
@@ -0,0 +1,1123 @@
...
+            # Note: there is no need to pad parameter matrices here, as there is just one new token
+            # for batched generation
+            dt = dt.unsqueeze(1) if dt.ndim == 2 else dt[:, 0, :].unsqueeze(1)
+            dt = dt.transpose(1, 2).expand(dt.shape[0], dt.shape[-1], self.head_dim)
+            # [num_heads] -> [num_heads, head_dim]
+            dt_bias = self.dt_bias.unsqueeze(-1).expand(self.dt_bias.shape[0], self.head_dim)
+
+            dt = torch.nn.functional.softplus(dt + dt_bias.to(dt.dtype))
+            dt = torch.clamp(dt, self.time_step_min) #, self.time_step_max)
+            A = A[..., None, None].expand(A.shape[0], self.head_dim, self.ssm_state_size).to(dtype=torch.float32)
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:473]: shape[0] is batch size, reused a lot, let's define them, same for sjape[1] and shape[-1]

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 17]
```diff
@@ -0,0 +1,1123 @@
...
+            cache_params.ssm_states[self.layer_idx].copy_(
+                cache_params.ssm_states[self.layer_idx] * dA + dBx
+            )
+
+            # Subsequent output
+            # [bsz, n_groups * state_size] -> [bsz, num_heads, state_size]
+            C = C.reshape(C.shape[0], self.n_groups, -1).unsqueeze(-2)
+            C = C.expand(C.shape[0], C.shape[1], self.num_heads // self.n_groups, C.shape[-1]).contiguous()
+            C = C.reshape(C.shape[0], -1, C.shape[-1])
+            # [bsz, num_heads, head_dim]
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:501]: explicit shape will make this more readable + None expansion

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 18]
```diff
@@ -0,0 +1,1123 @@
...
+
+            # Step 3: Compute Y_diag (apply to values)
+            #Y_diag = ((M.unsqueeze(-1) * hidden_states.unsqueeze(1)).sum(dim=3))
+            #Y_diag_einsum = torch.einsum("bclsh,bcshp->bclhp", M, hidden_states)
+            # Y_diag_alt = (M.unsqueeze(-1) * hidden_states.unsqueeze(2)).sum(3)
+            #diff_ = ((Y_diag_alt - Y_diag_einsum) / (Y_diag_einsum + 1e-9)).min()
+
+            #Y_diag_intermediate = M[..., None] * hidden_states[:, None, ...]
+            #Y_diag = Y_diag_intermediate.sum(dim=3)
+
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:564]: to remove?

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 19]
```diff
@@ -0,0 +1,1123 @@
...
+            Y_diag = (M.unsqueeze(-1) * hidden_states.unsqueeze(2)).sum(3)
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:565]: None notationt!

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 20]
```diff
@@ -0,0 +1,1123 @@
...
+            states = (B_decay_contraction.permute(0, 1, 3, 2, 4)[..., None]  * hidden_states.permute(0, 1, 3, 2, 4)[..., None, :]).sum(dim=3).permute(0, 1, 2, 4, 3)
+            if cache_params is not None and cache_params.seqlen_offset > 0:
+                previous_states = cache_params.ssm_states[self.layer_idx].unsqueeze(1)
+            else:
+                previous_states = torch.zeros_like(states[:, :1])
+            states = torch.cat([previous_states, states], dim=1)
+            decay_chunk = torch.exp(segsum(nn.functional.pad(A_cumsum[:, :, :, -1], (1, 0))))
+
+            states_permuted = states.permute(0, 2, 1, 3, 4)
+            result = (decay_chunk[..., None, None] * states_permuted[:, :, None, ...]).sum(dim=2)
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:581]: permuting decay chunk could avoid 2 permutes? (just wondering but ignore this comment!)

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 21]
```diff
@@ -0,0 +1,1123 @@
...
+
+            # D skip connection
+            # [num_heads] -> [num_heads, head_dim]
+            D = self.D.unsqueeze(-1).expand(self.D.shape[0], self.head_dim)
+            y = (y + hidden_states * D).to(y.dtype)
+
+            # [bsz, num_heads, head_dim] -> [bsz, 1, intermediate_size]
+            y = y.reshape(y.shape[0], -1).unsqueeze(1)
+        else:
+            # begin ssd naive implementation without einsums
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:518]: wow that's heavy thanks for figuring it out!

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 22]
```diff
@@ -0,0 +1,1123 @@
...
+    ) -> Dict[str, Any]:
+        model_kwargs["cache_params"] = outputs.get("cache_params", None)
+        if (
+            model_kwargs.get("use_cache", True)
+            and "cache_position" in model_kwargs
+            and model_kwargs["cache_position"] is not None
+        ):
+            model_kwargs["cache_position"] = model_kwargs["cache_position"][-1:] + num_new_tokens
+
+        return model_kwargs
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:996]: if you are not using them for now, ff to remove

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 23]
```diff
@@ -0,0 +1,1123 @@
...
+
+        model_inputs.update(
+            {
+                "attention_mask": attention_mask,
+                "cache_params": cache_params,
+                "use_cache": use_cache,
+                "cache_position": cache_position,
+            }
+        )
+        return model_inputs
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/modeling_mamba2.py:1060]: we need this only if you use the cache positions

**Review by ArthurZucker — COMMENTED** [review]
Congrats! There are a few nits here and there but LGTM otherwise!

**Comment by ArthurZucker** [comment]
The `.md` file should include your findings about speed performances etc

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 23 above)
Comment (molbap) [reply src/transformers/models/mamba2/modeling_mamba2.py:1060]: We need it too for the attention_mask, right?

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 14 above)
Comment (molbap) [reply src/transformers/models/mamba2/modeling_mamba2.py:374]: I think I got rid of all generate bugs in the end, they were related to the attention mask slicing!

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 15 above)
Comment (molbap) [reply src/transformers/models/mamba2/modeling_mamba2.py:399]: it's because of the convolution: even if we tune out the hidden states before it, we need to tune the remainder again to 0 afterwards!

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 5 above)
Comment (molbap) [reply src/transformers/models/mamba2/modeling_mamba2.py]: woops, missed it

**Comment by molbap** [comment]
Addressed comments and fixed some more tests, also tried a simple PEFT script (almost same as mamba 1) which seems to run :grinning: added it to the .md.

**Comment by molbap** [comment]
cc @ArthurZucker , the only test failing is the slow test because runner doesn't have access

File: docs/source/en/model_doc/mamba2.md [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
...`torch_forward` or `cuda_kernels_forward`. The latter uses the original cuda kernels if they are found in your environment, and is slower on the prefill i.e. requires a "warmup run" due to high cpu overhead, see [here](https://github.com/state-spaces/mamba/issues/389#issuecomment-2171755306) and [also here](https://github.com/state-spaces/mamba/issues/355#issuecomment-2147597457). Without compilation, the `torch_forward` implementation is faster by a factor 3 to 4. Further, there are no positional embeddings in this model, but there is an `attention_mask` and a specific logic to mask out hidden states in two places in the case of batched generation, see [here](https://github.com/state-spaces/mamba/issues/66#issuecomment-1863563829) as well. Due to this, in addition to the reimplementation of mamba2 kernels, batched generation and cached generation are expected to have slight discrepancies. Further, the results given by the cuda kernels or the torch forward are expected to be slightly different. The SSM algorithm heavily relies on tensor contractions, which have matmul equivalents but the order of operations is slightly different, making the difference greater at smaller precisions.
```
Comment (vasqu) [inline docs/source/en/model_doc/mamba2.md:31]: Small nit: Can we add left padding to the tips so people avoid using right padding (which likely doesn't work as expected) :eyes:

Comment (molbap) [reply docs/source/en/model_doc/mamba2.md:31]: It's by default activated, but you're right, better mention it here!

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 25]
```diff
@@ -0,0 +1,762 @@
...
+        self.mixer = Mamba2Mixer(config, layer_idx=layer_idx)
+
+    def forward(self, hidden_states, cache_params: Optional[Mamba2Cache] = None):
+        residual = hidden_states
+        hidden_states = self.norm(hidden_states.to(dtype=self.norm.weight.dtype))
+        if self.residual_in_fp32:
+            residual = residual.to(torch.float32)
+
+        hidden_states = self.mixer(hidden_states, cache_params=cache_params)
+        hidden_states = residual + hidden_states
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/mamba2/modeling_mamba2.py:651]: I think I've had it wrong here since the norm takes care of the downcast... mb

It will lead to fp32 in the last output of the block though which might be avoided by a downcast. For example, a bf16 input is computed correctly but the last output given to the rms norm ("norm_f") will take in fp32 and subsequently gets fp32 again. This leads to dtype mismatches if we expect a bf16 output for other subsequent components. 

In short, calculations for fp16/bf16 are correct but we get fp32 as the final output which shouldn't be the case imo.

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 14 above)
Comment (ArthurZucker) [reply src/transformers/models/mamba2/modeling_mamba2.py]: cool!

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/mamba2/modeling_mamba2.py:651]: cc @molbap not sure about this one but would love to hear your opinion on this.

**Comment by vasqu** [comment]
Looking really great! :fire: Thanks for the work

File: docs/source/en/model_doc/mamba2.md [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+trainer = SFTTrainer(
+    model=model,
+    tokenizer=tokenizer,
+    args=training_args,
+    peft_config=lora_config,
+    train_dataset=dataset,
+    dataset_text_field="quote",
+)
+trainer.train()
+## Mamba2Config
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/mamba2.md:85]:
```suggestion
trainer.train()
```

## Mamba2Config
```

**Review by ArthurZucker — APPROVED** [review]
LGTM, let's fix CI and we can merge!

File: docs/source/en/model_doc/mamba2.md (hunk 26 above)
Comment (molbap) [reply docs/source/en/model_doc/mamba2.md]: ah, just fixed that in the last README update!

File: docs/source/en/model_doc/mamba2.md [hunk 27]
```diff
@@ -29,6 +29,7 @@ Tips:
...
+Another note, shutdown of hidden states corresponding to padding tokens is done in 2 places and mostly has been tested with left-padding. Right-padding will propagate noise down the line and is not guaranteed to yield satisfactory results. `tokenizer.padding_side = "left"` ensures you are using the correct padding side.
 
 This model was contributed by [Molbap](https://huggingface.co/Molbap), with tremendous help from [Anton Vlasjuk](https://github.com/vasqu).
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/mamba2.md:34]: cc @vasqu thanks for your input in this PR! 🥳

Comment (molbap) [reply docs/source/en/model_doc/mamba2.md:34]: yes thanks a lot @vasqu , glad to have had your insights along the way, it was incredibly helpful!

**Comment by Adibvafa** [comment]
Super excited for this merge!
@vasqu Should we close https://github.com/huggingface/transformers/pull/32027 now?

**Comment by vasqu** [comment]
@Adibvafa Sry, it kinda drowned in my notifications! I'll close it for now, but if needed, it can also be reopened at any time.
